### PR TITLE
DDBTEAM-990: Added PressReader vendor service

### PR DESCRIPTION
**Pull request by cableman**
#### Link to ticket

https://jira.itkdev.dk/browse/DDBTEAM-990
https://platform.dandigbib.org/issues/5358

#### Description

Adding new vendor to fetch covers from press readers CDN.

#### Screenshot of the result

Screenshot 2022-04-04 at 16 54 06

#### Checklist

- [x] My code is covered by test cases.
- [x] My code passes our test (all our tests).
- [ ] My code passes our static analysis suite.
- [x] My code passes our continuous integration process.

If your code does not pass all the requirements on the checklist you have to add a comment explaining why this change 
should be exempt from the list.

#### Additional comments or questions

No comments

### CODE REVIEW

**Comment by cableman** [comment]
Please ignore psalm has there is another https://github.com/danskernesdigitalebibliotek/ddb-cover-service-importers/pull/109 about this types thing.

File: src/Service/VendorService/PressReader/PressReaderVendorService.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+{
+    use ProgressBarTrait;
+    use VendorServiceTrait;
+
+    protected const VENDOR_ID = 19;
+    private const VENDOR_ARCHIVE_NAME = 'pressreader';
+
+    private DataWellSearchService $datawell;
+
+    private string $urlPattern = 'https://i.prcdn.co/img?cid=%s&page=1&width=1200';
```
Comment (turegjorup) [inline src/Service/VendorService/PressReader/PressReaderVendorService.php:32]: All magic values should be class constants

File: src/Service/VendorService/PressReader/PressReaderVendorService.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    private VendorImageValidatorService $imageValidatorService;
+
+    /**
+     * DataWellVendorService constructor.
```
Comment (turegjorup) [inline src/Service/VendorService/PressReader/PressReaderVendorService.php:36]: PressReaderVendorService constructor

File: src/Service/VendorService/PressReader/PressReaderVendorService.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        $this->loadConfig();
+
+        $status = new VendorStatus();
+
+        $this->progressStart('Search data well for: "'.self::VENDOR_ARCHIVE_NAME.'"');
+
+        $offset = 1;
+        try {
+            do {
+                // Search the data well for material with acSource set to "comics plus".
```
Comment (turegjorup) [inline src/Service/VendorService/PressReader/PressReaderVendorService.php:66]: comics plus?

File: src/Service/VendorService/PressReader/PressReaderVendorService.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                [$pidArray, $more, $offset] = $this->datawell->search(self::VENDOR_ARCHIVE_NAME, $offset);
+                $this->TransformUrls($pidArray);
+
+                // The press reader CDN insert at special image saying that the content is not updated for newest news
+                // cover. See https://i.prcdn.co/img?cid=9L09&page=1&width=1200, but the size will be under 30Kb, so we have
+                // this extra test.
```
Comment (turegjorup) [inline src/Service/VendorService/PressReader/PressReaderVendorService.php:72]: 👍

File: src/Service/VendorService/PressReader/PressReaderVendorService.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                $pidArray = array_filter($pidArray, function ($url) {
+                    $header = $this->imageValidatorService->remoteImageHeader('cf-polished', $url);
+                    if (!empty($header)) {
+                        $header = reset($header);
+                        list($label, $size) = explode('=', $header);
+                        if ($size < 30000) {
```
Comment (turegjorup) [inline src/Service/VendorService/PressReader/PressReaderVendorService.php:78]: Comment is good but consider making the magic 30000 number a class constant

File: src/Service/VendorService/PressReader/PressReaderVendorService.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        $vendor = $this->vendorCoreService->getVendor($this->getVendorId());
+
+        // Set the service access configuration from the vendor.
+        $this->datawell->setSearchUrl($vendor->getDataServerURI());
+        $this->datawell->setUser($vendor->getDataServerUser());
+        $this->datawell->setPassword($vendor->getDataServerPassword());
+    }
+
+    /**
+     * Transform the URL from the datawell to CDN https://i.prcdn.co/img?cid={$id}&page=1&width=1200.
```
Comment (turegjorup) [inline src/Service/VendorService/PressReader/PressReaderVendorService.php:123]: Adding an example of the URLs from the datawell will increase readability

**Review by turegjorup — APPROVED** [review]
Approved with comments

File: src/Service/VendorService/PressReader/PressReaderVendorService.php (hunk 6 above)
Comment (cableman) [reply src/Service/VendorService/PressReader/PressReaderVendorService.php:123]: Fixed

File: src/Service/VendorService/PressReader/PressReaderVendorService.php (hunk 5 above)
Comment (cableman) [reply src/Service/VendorService/PressReader/PressReaderVendorService.php:78]: Fixed

File: src/Service/VendorService/PressReader/PressReaderVendorService.php (hunk 4 above)
Comment (cableman) [reply src/Service/VendorService/PressReader/PressReaderVendorService.php:72]: Fixed

File: src/Service/VendorService/PressReader/PressReaderVendorService.php (hunk 3 above)
Comment (cableman) [reply src/Service/VendorService/PressReader/PressReaderVendorService.php:66]: Fixed

File: src/Service/VendorService/PressReader/PressReaderVendorService.php (hunk 2 above)
Comment (cableman) [reply src/Service/VendorService/PressReader/PressReaderVendorService.php:36]: Fixed

File: src/Service/VendorService/PressReader/PressReaderVendorService.php (hunk 1 above)
Comment (cableman) [reply src/Service/VendorService/PressReader/PressReaderVendorService.php:32]: Fixed

**Pull request merged**